### PR TITLE
Keycodes: improve tree shaking by annotating exports as pure

### PR DESCRIPTION
**Pull request by jsnajdr**
The `keycodes` package has a few exports whose values are result of a function call:
```js
export const rawShortcut = mapValues( modifiers, createShortcut );
```
The problem with this is that bundlers like esbuild cannot tree shake this code when the `rawShortcut` export is unused. Because the bundler doesn't know if `mapValues` has any side effects.

The [esbuild docs](https://esbuild.github.io/api/#tree-shaking-and-side-effects) recommend using a `/* @__PURE__ */` comment:
```js
export const rawShortcut = /* @__PURE__ */ mapValues( modifiers, createShortcut );
```
They also recommend using an IIFE but in our case the expression is already a `mapValues` function call, so we don't need to wrap it.

The packages has the `sideEffects: false` flag in `package.json` but it doesn't help here.

**How to test:**
This is very visible in the `packages/dataviews/build-wp/index.js` bundle. The only thing that the bundle uses from `keycodes` is the `ESCAPE` constant. But the bundle contains also all the code with possible side effects.

After this PR, the definition of `ESCAPE` is all that remains from `keycodes`.

It would be nice to also test this with webpack, but I didn't do that yet.

### CODE REVIEW

File: packages/keycodes/src/index.js [hunk 1]
```diff
@@ -211,9 +211,9 @@ export const modifiers = {
  * @type {WPModifierHandler<WPKeyHandler<string>>} Keyed map of functions to raw
  *                                                 shortcuts.
  */
-export const rawShortcut = mapValues(
-	modifiers,
-	( /** @type {WPModifier} */ modifier ) => {
+export const rawShortcut =
+	/* @__PURE__ */
```
Comment (youknowriad) [inline packages/keycodes/src/index.js:215]: Is this something supported by all bundlers?

Comment (tyxla) [reply packages/keycodes/src/index.js:215]: Not sure if it's supported by all, but generally, most would support it, as it's declared in the spec:

https://github.com/javascript-compiler-hints/compiler-notations-spec/blob/main/pure-notation-spec.md

also see compatibility table:

https://github.com/javascript-compiler-hints/compiler-notations-spec/blob/main/pure-notation-compatibility.md

**Review by tyxla — COMMENTED** [review]
This should do the job, but I wonder if we can instruct webpack and other bundlers to always consider `mapValues()` as a pure function call. That way, we won't have to manually do it everywhere.

File: packages/keycodes/src/index.js (hunk 1 above)
Comment (jsnajdr) [reply packages/keycodes/src/index.js:215]: The webpack + Terser combo also supports it. Webpack adds the comment to code that is pure (typically re-exports) and then Terser uses the comment to eliminate the dead code.

**Comment by jsnajdr** [comment]
> I wonder if we can instruct webpack and other bundlers to always consider `mapValues()` as a pure function call.

I never heard about such a feature or plugin. Also, it's not enough that `mapValues` is pure, the callback also needs to be pure. And that's hard to verify.

**Comment by sgomes** [comment]
> This should do the job, but I wonder if we can instruct webpack and other bundlers to always consider `mapValues()` as a pure function call. That way, we won't have to manually do it everywhere.

`/* @__PURE__ */` only works on call expressions, so there's no direct way of doing this.

That said, I believe another approach to work around the issue would be to move each of these exports into its own module, and then re-export them all from this one. Since we already have the `sideEffects: false` flag in `package.json`, there should be no additional changes needed. I could be wrong, though; there may be a deeper inspection of side effects going on than what I expect.

**Comment by jsnajdr** [comment]
> another approach to work around the issue would be to move each of these exports into its own module

Yes, that should also solve the problem, but it's a lot of work, there are many modules and they are tiny and interdependent. We could connect that with a TypeScript rewrite.

**Review by youknowriad — APPROVED** [review]

**Pull request merged**